src/Microsoft.Health.Fhir.Core/Features/Search/Expressions/Parsers/ExpressionParser.cs

```diff
@@ -101,7 +102,14 @@ public IncludeExpression ParseInclude(string resourceType, string includeValue,
                 refSearchParameter = _searchParameterDefinitionManager.GetSearchParameter(originalType.ToString(), searchParam.ToString());
             }
 
-            return new IncludeExpression(resourceType, refSearchParameter, originalType.ToString(), targetType, wildCard, isReversed);
+            referencedTypes = new List<string>();
+            _searchParameterDefinitionManager.GetSearchParameters(resourceType)
+            .Where(p => p.Type == ValueSets.SearchParamType.Reference).ToList()
```

Comment:
This looks pretty allocation-heavy. Also, it seems like something we could cache (and populate lazily).

Reply:
Agree. The challenge was to propagate the searchParameterDefinitionManager to the IncludeExpression, so that it will be used upon need. That didn't make much sense to me since IncludeExpression represents a parsed expression, so ideally searchParameterDefinitionManager shouldn't be used in it. 

Maybe we can use a delegate that is applied once when ReferencedTypes is used and populates it. WDYT?

Reply:
Actually, after giving it another thought, It’s worth noting that it will always be used when the expression is _include wildcard followed by an /include iterate (or revinclude), since I am storing the cte names in the dictionary.

So I suggest to always populate it for wildcard followed by  iterate expression.
So basically the search options factory “knows” if there are include iterate expressions.

WDYT?